marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java

```diff
@@ -112,8 +113,9 @@ public FlowRunnerImpl(String host, String username, String password) {
      */
     public FlowRunnerImpl(HubConfig hubConfig) {
         this.hubConfig = hubConfig;
-        this.stepRunnerFactory = new StepRunnerFactory(this.hubConfig);
-        this.stepRunnerFactory.setStepDefinitionProvider(new MarkLogicStepDefinitionProvider(this.hubConfig.newStagingClient(null)));
+        this.flowManager = new FlowManagerImpl(hubConfig);
+        this.stepRunnerFactory = new StepRunnerFactory(hubConfig);
```

Comment:
My gut feeling is the StepRunnerFactory can be implemented as a singleton object.   StepRunnerFactory does not need to keep those states' reference when getStepRunner gets called. MarkLogicStepDefinitionProvider can be implemented as a singleton object. I am not sure why we have to pass stepRunnerFactory to the Runnable?

Reply:
The StepDefinitionProvider can't easily be a singleton because we have multiple implementations of it. As for FlowRunnerTask depending on StepRunnerFactory - it's likely that whole design can change, as I don't know that a FlowRunnerImpl is ever really used to run multiple flows at once. But I don't think it's worth changing anything about that right now.

Reply:
@rjudin I do not think I need to do now. But I think I can refractor for sure. I even think that we should be able to run multiple flows concurrently with a better threadpool (i/o operations) management in backend.

Reply:
Currently we don't have any requirement for running multiple flows though the same FlowRunnerImpl instance, so let's not tackle this until such a requirement comes up. When it does, I'd rather see a FlowRunnerImpl with far less state in it, such that we can spin up a new instance easily to run a new flow.

Reply:
yes, I thought that maybe we did not have use cases to do that in the past. But I do not know the reason.